### PR TITLE
Issue #6104: fix GenericWhitespaceCheck for ctor calls

### PR DESCRIPTION
**Pull request by strkkk**
#6104 

Report https://strkkk.github.io/checkstyle/6104/index.html
Report after update https://strkkk.github.io/checkstyle/6104_2/index.html

### CODE REVIEW

**Comment by pbludov** [comment]
The report is good.

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/genericwhitespace/InputGenericWhitespaceDefault.java [hunk 1]
```diff
@@ -77,5 +77,9 @@ public void beforeAndAfter() {
         Map<Class<?>, Integer> a = new HashMap<Class<?>, Integer>();
         Map<Class<?>, Integer> b = (Map<Class<?>, Integer>) a;
     }
+    Object ok = new <String>Object();
+    Object notOkStart = new<String>Object();
```
Comment (pbludov) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/genericwhitespace/InputGenericWhitespaceDefault.java]: Please mark these lines with the `// violation` comment.

Comment (strkkk) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/genericwhitespace/InputGenericWhitespaceDefault.java]: done

**Comment by strkkk** [comment]
Appveyor failed because of some infra issue

**Comment by romani** [comment]
Strange failure:
`Failed to execute goal org.apache.maven.plugins:maven-site-plugin:3.8.2:site (default-site) on project checkstyle: Error parsing 'C:\projects\checkstyle\src\xdocs\anttask.xml.vm': line [-1] Error validating the model: Connect to maven.apache.org:80 [maven.apache.org/95.216.24.32, maven.apache.org/40.79.78.1] failed: Connection timed out: connect -> [Help 1]`

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/GenericWhitespaceCheck.java [hunk 2]
```diff
@@ -229,6 +229,16 @@ else if (!isCharacterValidAfterGenericEnd(charAfter)) {
...
     }
 
+    /**
+     * Checks if generic is before constructor invocation.
+     * @param ast ast
+     * @return true if generic before a constructor invocation
+     */
+    private static boolean isGenericBeforeCtor(DetailAST ast) {
+        return ast.getParent().getParent().getType() == TokenTypes.LITERAL_NEW
+                && ast.getParent().getNextSibling().getType() == TokenTypes.IDENT;
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/GenericWhitespaceCheck.java]: When is it not identifier?
And does this cover things like `new OuterClass.InnerClass()` or `new package.Class()`.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/GenericWhitespaceCheck.java [hunk 3]
```diff
@@ -219,7 +219,7 @@ private void processSingleGeneric(DetailAST ast, String line, int after) {
         //    Collections.<Object>emptySet();
         //                        ^
         //                        +--- whitespace not allowed
-        if (isGenericBeforeMethod(ast)) {
+        if (isGenericBeforeMethod(ast) || isGenericBeforeCtor(ast)) {
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/GenericWhitespaceCheck.java]: Please remove upper comment completely.
It now becomes a bit misleading.

Please update doc https://checkstyle.org/config_whitespace.html#GenericWhitespace at description to make it
```
- should not be preceded with whitespace when it is precede method name or constructor
- should not be preceded with whitespace when following type name.
```

And 
```
should be followed with whitespace in almost all cases, except diamond operators and when preceding method name or contructor.
```

Please add example of correct ctor to `Examples with correct spacing:` under example for method

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/GenericWhitespaceCheck.java (hunk 2 above)
Comment (strkkk) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/GenericWhitespaceCheck.java]: It is not identifier in cases like new ArrayList<String>

Comment (strkkk) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/GenericWhitespaceCheck.java]: Done
I added test cases and code changes to detect static classes and package

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/GenericWhitespaceCheck.java (hunk 3 above)
Comment (strkkk) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/GenericWhitespaceCheck.java]: Done

**Comment by strkkk** [comment]
@rnveach @romani I corrected review items, also updated report (spoiler - it is exactly same as first one).
Please take a look.

**Pull request merged**